src/plugin_pathtranslator.cpp

```diff
@@ -0,0 +1,197 @@
+    JASSERT(colonIdx != std::string::npos)(token).Text("Bad mapping; expect old:new");
+    string oldPath = token.substr(0, colonIdx);
+    string newPath = token.substr(colonIdx + 1);
+    JASSERT(!oldPath.empty());
+    JASSERT(!newPath.empty());
+    (*pathMapping)[oldPath] = newPath;
+  }
+}
+
+static void pathTranslator_Init()
+{
+  if (pathMapping == nullptr) {
+    pathMapping = new(pathMappingStorage) unordered_map<string, string>;
+    populatePathMapping(getenv(ENV_PATH_MAPPING));
+  }
+}
+
+static void pathTranslator_Restart()
+{
+  pathTranslator_Init();
+  char *tmp = (char*) JALLOC_MALLOC(MAX_ENV_VAR_SIZE);
+  DmtcpGetRestartEnvErr_t ret = dmtcp_get_restart_env(ENV_PATH_MAPPING, tmp, MAX_ENV_VAR_SIZE);
+  if (ret == RESTART_ENV_SUCCESS) {
+    populatePathMapping(tmp);
+  }
+
+  JALLOC_FREE(tmp);
+}
+
+static void
+pathTranslator_PrepareForExec(DmtcpEventData_t *data)
+{
+  pathTranslator_Init();
+  JASSERT(data != NULL);
+  jalib::JBinarySerializeWriterRaw wr("", data->preExec.serializationFd);
+  wr.serialize(*pathMapping);
+}
+
+static void
+pathTranslator_PostExec(DmtcpEventData_t *data)
+{
+  pathTranslator_Init();
+  JASSERT(data != NULL);
+  jalib::JBinarySerializeReaderRaw rd("", data->postExec.serializationFd);
+  rd.serialize(*pathMapping);
+}
+
+/*
+ * virtual_to_physical_path - translator virtual to physical path
+ *
+ * Returns a string for the corresponding physical path to the given
+ * virtual path. If no path translation occurred, the given virtual path
+ * will simply be returned as a string.
+ *
+ * Conceptually, an original path prior to the first checkpoint is considered a
+ * "virtual path".  After a restart, it will be substituted using the latest
+ * list of registered paths.  Hence, a newly registered path to be substituted
+ * is a "physical path".  Internally, DMTCP works with the original "virtual
+ * path" as the canonical name.  But in any system calls, it must translator the
+ * virtual path to the latest "physical path", which will correspond to the
+ * current, post-restart filesystem.
+ */
+
+static void
+pathTranslator_VirtualToReal(DmtcpEventData_t *data)
+{
+  pathTranslator_Init();
+  char *virtPath = data->virtualToRealPath.path;
+
+  // Check to see if prefix of virtPath is present in pathMappings.
+  // This is O(n) in the number of mappings. We could make this faster by using
+  // a trie or a hash table.
+  // TODO (kapil): Investigate if we need to match the longest prefix instead of
+  // the first one.
+  for (const auto &mapping : *pathMapping) {
+    const string &oldp = mapping.first;
+    const string &newp = mapping.second;
+    if (Util::strStartsWith(virtPath, oldp.c_str())) {
+      // boundary check: next char must be '/' or end
+      char next = virtPath[oldp.size()];
+      if (next == '/' || next == '\0') {
+        const char* suffix = virtPath + oldp.size();
+        int n = snprintf(data->virtualToRealPath.path, PATH_MAX, "%s%s", newp.c_str(), suffix);
```

Comment:
Maybe we can add some comments here, indicating we are modifying the `data` as the return value.